File: src/main/java/org/gd/leetcode/p0814/Solution.java
package org.gd.leetcode.p0814;

import org.gd.leetcode.common.LeetCode;
import org.gd.leetcode.common.TreeNode;

/**
 * https://leetcode.com/problems/binary-tree-pruning/
 *
 * @author Horkhover Dmytro
 * @since 2018-12-18
 */
@LeetCode(difficulty = LeetCode.Level.MEDIUM, tags = LeetCode.Tags.TREE)
class Solution {

    private static void pruneLeft(TreeNode parent, TreeNode node) {
        if (node != null) {
            pruneLeft(node, node.left);
            pruneRight(node, node.right);
            if (node.left == null && node.right == null && node.val == 0)
                parent.left = null;
        }
    }

    private static void pruneRight(TreeNode parent, TreeNode node) {
        if (node != null) {
            pruneLeft(node, node.left);
            pruneRight(node, node.right);
            if (node.left == null && node.right == null && node.val == 0)
                parent.right = null;
        }
    }

    public TreeNode pruneTree(TreeNode root) {
        if (root != null) {
            pruneLeft(root, root.left);
            pruneRight(root, root.right);
        }
        return root;
    }
}

File: src/main/java/org/gd/leetcode/p0814/Solution.py
# https://leetcode.com/problems/binary-tree-pruning/

class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    
    def pruneTree(self, root: TreeNode) -> TreeNode:
        if root:
            self.prune_left(root, root.left)
            self.prune_right(root, root.right)

        return root

    def prune_left(self, parent: TreeNode, node: TreeNode):

        if node:
            self.prune_left(node, node.left)
            self.prune_right(node, node.right)
            
            if not node.left and not node.right and node.val == 0:
                parent.left = None

    def prune_right(self, parent: TreeNode, node: TreeNode):

        if node:
            self.prune_left(node, node.left)
            self.prune_right(node, node.right)
            
            if not node.left and not node.right and node.val == 0:
                parent.right = None

